src/abstractgraphindices.jl

```diff
@@ -0,0 +1,40 @@
+using Graphs: AbstractEdge
+
+abstract type AbstractGraphIndices{T} <: AbstractVector{T} end
+abstract type AbstractVertices{V} <: AbstractGraphIndices{V} end
+abstract type AbstractEdges{V, E <: AbstractEdge{V}} <: AbstractGraphIndices{E} end
+
+struct Vertices{V} <: AbstractVertices{V}
+    vertices::Vector{V}
+end
+struct Edges{V, E} <: AbstractEdges{V, E}
+    edges::Vector{E}
+    Edges(edges::Vector{E}) where {V, E <: AbstractEdge{V}} = new{V, E}(edges)
+end
```

Comment:
I think we should make the collections more generic:
```suggestion
struct Vertices{V, Vs <: AbstractVector{V}} <: AbstractVertices{V}
    vertices::Vs
end
struct Edges{V, E <: AbstractEdge{V}, Es <: AbstractVector{E}} <: AbstractEdges{V, E}
    edges::Es
end
```
so then they just act as lightweight wrappers around whatever collection the vertices or edges are in.

Reply:
I guess we can be even more generic and only assume that the collection implements `eltype`:
```Julia
struct Vertices{V, Vs} <: AbstractVertices{V}
    vertices::Vs
    Vertices(vertices::Vs) where {Vs} = new{eltype{Vs}, Vs}(vertices)
end
struct Edges{V, E <: AbstractEdge{V}, Es} <: AbstractEdges{V, E}
    edges::Es
    function Edges(edges::Es) where {Es}
        E = eltype(Es)
        return new{vertextype{E}, E, Es}(edges)
    end
end
```

Reply:
`Indices` is a good example, though the issue there is that `getindex` acts in a different way for `Indices` so that is kind of an awkward case (i.e. if `Vertices` is wrapping `Indices`, `Base.getindex(vs::Vertices, i::Int) = vs.vertice[i]` would be wrong). I'm not entirely sure, but my feeling would be that it if `Vertices` is an `AbstractVector`, it should wrap an `AbstractVector` so then you know you can safely forward `AbstractVector` functions, but otherwise if we want to wrap any iterable over vertices it shouldn't be an `AbstractVector` (i.e. it's subtyping should reflect what it is allowed to wrap).

Reply:
One question is, is there a reason for `Vertices` to have linear indexing? If not it may as well not be an `AbstractVector` and instead just be an iterator. That seems reasonable since `vertices(graph)` and `edges(graph)` generally output iterators anyway.